pkg/gofr/datasource/dgraph/draph.go

```diff
@@ -0,0 +1,251 @@
+
+	dg := dgo.NewDgraphClient(api.NewDgraphClient(conn))
+	d.client = NewDgraphClient(dg)
+
+	// Check connection by performing a basic health check
+	if _, err := d.HealthCheck(context.Background()); err != nil {
+		d.logger.Errorf("dgraph health check failed: %v", err)
+		return
+	}
+}
+
+// UseLogger sets the logger for the Dgraph client which asserts the Logger interface.
+func (d *Client) UseLogger(logger any) {
+	if l, ok := logger.(Logger); ok {
+		d.logger = l
+	}
+}
+
+// UseMetrics sets the metrics for the Dgraph client which asserts the Metrics interface.
+func (d *Client) UseMetrics(metrics any) {
+	if m, ok := metrics.(Metrics); ok {
+		d.metrics = m
+	}
+}
+
+// Query executes a read-only query in the Dgraph database and returns the result.
+func (d *Client) Query(ctx context.Context, query string) (any, error) {
+	start := time.Now()
+
+	// Execute query
+	resp, err := d.client.NewTxn().Query(ctx, query)
+	duration := time.Since(start).Microseconds()
+
+	// Create and log the query details
+	ql := &QueryLog{
+		Type:     "Query",
+		URL:      query,
+		Duration: duration,
+	}
+
+	d.logger.Debug("executing dgraph query")
+	ql.PrettyPrint(d.logger)
+
+	if err != nil {
+		d.logger.Error("dgraph query failed: ", err)
+		return nil, err
+	}
+
+	d.logger.Debugf("dgraph query succeeded in %dµs", duration)
+
+	d.metrics.RecordHistogram(ctx, "dgraph_query_duration", float64(duration))
+
+	return resp, nil
+}
+
+// QueryWithVars executes a read-only query with variables in the Dgraph database.
+func (d *Client) QueryWithVars(ctx context.Context, query string, vars map[string]string) (any, error) {
+	start := time.Now()
+
+	// Execute the query with variables
+	resp, err := d.client.NewTxn().QueryWithVars(ctx, query, vars)
+	duration := time.Since(start).Microseconds()
+
+	// Create and log the query details
+	ql := &QueryLog{
+		Type:     "QueryWithVars",
+		URL:      fmt.Sprintf("Query: %s, Vars: %v", query, vars),
+		Duration: duration,
+	}
+
+	if err != nil {
+		d.logger.Error("dgraph queryWithVars failed: ", err)
+		ql.PrettyPrint(d.logger)
+		return nil, err
+	}
+
+	d.logger.Debugf("dgraph queryWithVars succeeded in %dµs", duration)
+	ql.PrettyPrint(d.logger)
+
+	d.metrics.RecordHistogram(ctx, "dgraph_query_with_vars_duration", float64(duration))
+
+	return resp, nil
+}
+
+// Mutate executes a write operation (mutation) in the Dgraph database and returns the result.
+func (d *Client) Mutate(ctx context.Context, mu any) (any, error) {
+	start := time.Now()
+
+	// Cast to proper mutation type
+	mutation, ok := mu.(*api.Mutation)
+	if !ok {
+		return nil, errInvalidMutation
+	}
+
+	// Execute mutation
+	resp, err := d.client.NewTxn().Mutate(ctx, mutation)
+	duration := time.Since(start).Microseconds()
+
+	// Create and log the mutation details
+	ql := &QueryLog{
+		Type:     "Mutation",
+		URL:      mutation.String(),
+		Duration: duration,
+	}
+	d.logger.Debug("executing dgraph mutation")
+	ql.PrettyPrint(d.logger)
```

Comment:
this can be kept in a function similar to how it is being done for other datasources (send operation stats) - it would make it look cleaner